content/300-guides/150-testing/150-integration-testing.mdx

```diff
@@ -1,13 +1,458 @@
+    productId: 1,
+    quantity: 1,
+  };
+
+  // Create the order and customer
+  await createOrder(order);
+
+  // Check if the new customer was created by filtering on unique email field
+  const newCustomer = await prisma.customer.findUnique({
+    where: {
+      email: customer.email,
+    },
+  });
+
+  // Check if the new order was created by filtering on unique email field of the customer
+  const newOrder = await prisma.order.findFirst({
+    where: {
+      customer: {
+        email: customer.email,
+      },
+    },
+  });
+
+  // Expect the new customer to have been created and match the input
+  expect(newCustomer).toEqual(customer);
+  // Expect the new order to have been created and contain the new customer
+  expect(newOrder).toHaveProperty("customerId", 2);
+
+  done()
+});
+
+it("should create 1 order with an existing customer", async (done) => {
+  // The existing customers email
+  const customer: Customer = {
+    email: "harry@hogwarts.io",
+  };
+  // The new orders details
+  const order: OrderInput = {
+    customer,
+    productId: 1,
+    quantity: 1,
+  };
+
+  // Create the order and connect the existing customer
+  await createOrder(order);
+
+  // Check if the new order was created by filtering on unique email field of the customer
+  const newOrder = await prisma.order.findFirst({
+    where: {
+      customer: {
+        email: customer.email,
+      },
+    },
+  });
+
+  // Expect the new order to have been created and contain the existing customer with an id of 1 (Harry Potter from the seed script)
+  expect(newOrder).toHaveProperty("customerId", 1);
+
+  done()
+});
+
+it("should show 'Out of stock' message if productId doesn't exit", async (done) => {
+  // The existing customers email
+  const customer: Customer = {
+    email: "harry@hogwarts.io",
+  };
+  // The new orders details
+  const order: OrderInput = {
+    customer,
+    productId: 3,
+    quantity: 1,
+  };
+
+  // The productId supplied doesnt exit so the function should return an "Out of stock" message
+  await expect(createOrder(order)).resolves.toEqual(new Error('Out of stock'));
+
+  done()
+});
+
+```
+
+## Running the tests
+
+This setup isolates a real world scenario so that you can test your applications functionality against real data in a controlled environment. 
+
+You can add some scripts to your projects `package.json` file which will setup the database and run the tests, then afterwards manually destroy the container.
+
+```json file=package.json
+  "scripts": {
+    "migrate": "npx prisma migrate dev --name init",
+    "docker:up": "docker-compose up -d",
+    "docker:down": "docker compose down",
+    "test": "yarn docker:up && yarn migrate && jest -i"
+  },
+```
+
+The `test` script does the following:
+
+1. Runs `docker compose up -d` to create the container with the Postgres image and database.
+1. Runs a migration on the database using your projects schema, this creates the tables in the containers database.
```

Comment:
```suggestion
1. Runs a migration on the database using your project's schema, this creates the tables in the container's database.
```

Reply:
Im unsure about this. For me the project is a possesive noun (it belongs to the user) so it doesn't need an apostrophe. The container should be changed though.

Ive forgotten alot of the rules, was a while ago i studied english! Will have a google around for more info. I could be wrong

Reply:
I went with "Runs a migration on the database using your projects schema, this creates the tables in the container's database."